### PR TITLE
[GEP-21] provider-local single-stack ipv6 implementation

### PR DESCRIPTION
**Pull request by nschad**
**How to categorize this PR?**
/area networking
/kind enhancement
/kind api-change

**What this PR does / why we need it**:

See https://github.com/gardener/gardener/pull/8574

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/7051

**Special notes for your reviewer**:

**Release note**:

```feature developer
add ipv6 support for this extension
```

### CODE REVIEW

File: pkg/apis/calico/types_network.go [hunk 1]
```diff
@@ -26,38 +26,51 @@ const (
...
+	PoolIPIP  Pool = "ipip"
+	PoolVXLan Pool = "vxlan"
 )
 
 // IPv4 contains configuration for calico ipv4 specific settings
 type IPv4 struct {
 	// Pool configures the type of ip pool for the tunnel interface.
 	// https://docs.projectcalico.org/v3.8/reference/node/configuration#environment-variables
-	Pool *IPv4Pool
+	Pool *Pool
```
Comment (axel7born) [inline pkg/apis/calico/types_network.go:51]: Do we really change the API here?

File: pkg/apis/calico/v1alpha1/types_network.go [hunk 2]
```diff
@@ -26,41 +26,59 @@ const (
...
 
+// IPv6 contains configuration for calico ipv6 specific settings
+type IPv6 struct {
+	// Pool configures the type of ip pool for the tunnel interface
+	// https://docs.tigera.io/calico/latest/reference/configure-calico-node#configuring-the-default-ip-pools
+	// +optional
+	Pool *Pool `json:"pool,omitempty"`
+	// Mode is the mode for the IPv6 Pool (e.g. Always, Never, CrossSubnet)
+	// TODO: VXLAN also supports CrossSubnet for VXLAN. Why is this not supported?
+	// vxlan pools accept only Always and Never (unchecked)
```
Comment (axel7born) [inline pkg/apis/calico/v1alpha1/types_network.go:73]: Though, I don't know the answer to the TODO question, I would prefer not to have it here in the api-reference.
You could open an issue regarding the CrossSubnet support?

File: pkg/apis/calico/types_network.go (hunk 1 above)
Comment (nschad) [reply pkg/apis/calico/types_network.go:51]: No it's basically just renamed. The only new thing for the API is the IPv6 stuff

File: pkg/apis/calico/v1alpha1/types_network.go (hunk 2 above)
Comment (nschad) [reply pkg/apis/calico/v1alpha1/types_network.go:73]: Absolutely the TODO makes no sense here. I will create an issue about it

File: pkg/apis/calico/types_network.go (hunk 1 above)
Comment (nschad) [reply pkg/apis/calico/types_network.go:51]: Ah okay I see. I figured in a IPV4 struct having the PoolMode to be named `IPv4PoolMode` is just word clutter. It also deduped some code and uses less "bloat".

However I don't mind reverting

Comment (timebertt) [reply pkg/apis/calico/types_network.go:51]: I would vote for keeping the short type names (without `IPv4`) as it fits the use across IPv4 and IPv6.
To keep the go API package backward-compatible, I suggest adding type aliases for `IPv4{Pool,PoolMode}`. That should be good enough :)

**Comment by timebertt** [comment]
PR is rebased and all comments addressed. PTAL @axel7born :)

File: pkg/apis/calico/types_network.go (hunk 1 above)
Comment (axel7born) [reply pkg/apis/calico/types_network.go:53]: OK, I'm fine with this.

File: pkg/controller/actuator_reconcile.go [hunk 3]
```diff
@@ -82,19 +84,29 @@ func (a *actuator) Reconcile(ctx context.Context, _ logr.Logger, network *extens
...
+			}
+			networkConfig.IPv6.AutoDetectionMethod = &autodetectionMode
 		}
-		networkConfig.IPv4.AutoDetectionMethod = &autodetectionMode
 	}
 
 	if networkConfig != nil {
 		if networkConfig.Overlay != nil && networkConfig.Overlay.Enabled {
-			networkConfig.IPv4.Mode = (*calicov1alpha1.IPv4PoolMode)(pointer.String(string(calicov1alpha1.Always)))
+			networkConfig.IPv4.Mode = (*calicov1alpha1.PoolMode)(pointer.String(string(calicov1alpha1.Always)))
```
Comment (DockToFuture) [inline pkg/controller/actuator_reconcile.go:105]:
```suggestion
			if ipFamilies.Has(extensionsv1alpha1.IPFamilyIPv4) {
				networkConfig.IPv4.Mode = (*calicov1alpha1.PoolMode)(pointer.String(string(calicov1alpha1.Always)))
			}
			if ipFamilies.Has(extensionsv1alpha1.IPFamilyIPv6) {
				networkConfig.IPv6.Mode = (*calicov1alpha1.PoolMode)(pointer.String(string(calicov1alpha1.Always)))
			}
```

File: pkg/controller/actuator_reconcile.go [hunk 4]
```diff
@@ -82,19 +84,29 @@ func (a *actuator) Reconcile(ctx context.Context, _ logr.Logger, network *extens
...
 			networkConfig.Backend = (*calicov1alpha1.Backend)(pointer.String(string(calicov1alpha1.Bird)))
 		}
 		if networkConfig.Overlay != nil && !networkConfig.Overlay.Enabled {
-			networkConfig.IPv4.Mode = (*calicov1alpha1.IPv4PoolMode)(pointer.String(string(calicov1alpha1.Never)))
+			networkConfig.IPv4.Mode = (*calicov1alpha1.PoolMode)(pointer.String(string(calicov1alpha1.Never)))
```
Comment (DockToFuture) [inline pkg/controller/actuator_reconcile.go:109]:
```suggestion
			if ipFamilies.Has(extensionsv1alpha1.IPFamilyIPv4) {
				networkConfig.IPv4.Mode = (*calicov1alpha1.PoolMode)(pointer.String(string(calicov1alpha1.Never)))
			}
			if ipFamilies.Has(extensionsv1alpha1.IPFamilyIPv6) {
				networkConfig.IPv6.Mode = (*calicov1alpha1.PoolMode)(pointer.String(string(calicov1alpha1.Never)))
			}
```

**Comment by nschad** [comment]
Hey @DockToFuture,

We have implemented the suggestions can you please PTAL?

**Comment by DockToFuture** [comment]
> Hey @DockToFuture,
> 
> We have implemented the suggestions can you please PTAL?

The PR looks good to me, we will test it once more, to validate that it does not break anything and then get it in. :)

File: pkg/controller/actuator_reconcile.go (hunk 4 above)
Comment (DockToFuture) [reply pkg/controller/actuator_reconcile.go:109]: Any reasons why this got now removed?

**Comment by DockToFuture** [comment]
> Hey @DockToFuture,
> 
> We have implemented the suggestions can you please PTAL?

Any reasons why this got now removed?

**Comment by nschad** [comment]
> > Hey @DockToFuture,
> > We have implemented the suggestions can you please PTAL?
> 
> Any reasons why this got now removed?

I messed up.

File: pkg/charts/utils.go [hunk 5]
```diff
@@ -204,10 +215,57 @@ func ComputeCalicoChartValues(
...
+			AutoDetectionMethod: nil,
+		}
+	}
+
+	if isIPv6 {
+		c.IPAM.AssignIPv6 = true
+		c.IPAM.Subnet = usePodCIDRv6
+		// NOTE: calico guide (https://docs.tigera.io/calico/latest/networking/ipam/ipv6#enable-ipv6-only) requires ipam
+		// type to be calico-ipam for single IPv6 or dual-stack clusters.
+		c.IPAM.IPAMType = calicoIPAM
```
Comment (nschad) [inline pkg/charts/utils.go:245]: Did we interpret this correctly? @DockToFuture

Comment (nschad) [reply pkg/charts/utils.go:245]: CC: @axel7born

**Comment by nschad** [comment]
Hey,

FYI: We dropped one PreFactor commit because it caused the E2E test to fail. We accidentally changed the behavior in this dropped commit from sometimes deploy `IPPool` to **always** deploy `IPPool` which causes the `IPPool` to be created by the extension with `IPIPMode: Never` which then causes the pod-network ping test to fail. 

The behavior is now back to how it was and the ippool is not deployed anymore  by the extension, instead it's created by calico itself.

File: pkg/charts/utils.go (hunk 5 above)
Comment (DockToFuture) [reply pkg/charts/utils.go:245]: Yes, unfortunately it's currently not possible to use the host-local IPAM plugin. I've raised an issue in regards to that: https://github.com/projectcalico/calico/issues/8190

**Review by DockToFuture — APPROVED** [review]
lgtm

**Pull request merged**